src/main/java/com/google/api/codegen/discovery/Schema.java

```diff
@@ -56,6 +61,64 @@ public Schema dereference() {
     return this;
   }
 
+  /**
+   * Traverses the schema's child nodes to find a Schema with the given childName. Returns the
+   * schema traversal path to the target; this path will include the starting node if the target was
+   * found. Returns an empty list if the target is not found.
+   */
+  public List<Schema> findChild(String childName) {
+    Set<Schema> visitedNodes = new HashSet<>();
+    Map<Schema, Schema> nodeToPrevNode = new HashMap<>();
+
+    Schema currentNode = this;
+    Queue<Schema> queue = new LinkedList<>();
+    queue.add(this);
+    visitedNodes.add(this);
+
+    // BFS to find the target node.
+    while (queue.size() != 0 && !currentNode.getIdentifier().equals(childName)) {
+      currentNode = queue.poll();
+
+      Queue<Schema> localQueue = new LinkedList<>();
+      if (!Strings.isNullOrEmpty(currentNode.reference())) {
+        localQueue.add(currentNode.dereference());
+      } else {
+        if (currentNode.properties() != null && currentNode.properties().size() > 0) {
+          localQueue.addAll(currentNode.properties().values());
+        }
+        if (currentNode.additionalProperties() != null) {
+          localQueue.add(currentNode.additionalProperties());
+        }
+      }
+
+      while (localQueue.peek() != null) {
+        Schema next = localQueue.poll();
+        nodeToPrevNode.put(next, currentNode);
```

Comment:
What if there are two or more paths to `next`, from different values of `currentNode`? This will overwrite any existing link from `next`.

If there can be no such link, does that mean this is a tree structure? If so, we don't need a full BFS.

Reply:
It's ok if an existing link gets overriden because we only need to find one path, not all paths, from the target back to the source. As long as there is at least one value of `currentNode` for every key `next`, we can trace the back along an unbroken path from the target to start.

Reply:
The schema is not guaranteed to be a tree.

Reply:
What a structure with a bidirectional link (or any cycle) like this, searching for `D` from `A`:
```
A-->B<-->C-->D
```

Won't it go like this:
- currentNode=`A`, add `B` to `localQueue`, add `(B->A)` to `nodeToPrevNode`, add `B` to `queue`
- currentNode=`B`, add `C` to `localQueue`, add `(C->B)` to `nodeToPrevNode`, add `C` to `queue`
- currentNode=`C`, add `B` and `D` to `localQueue`
  - Process `B` in `localQueue`, add `(B->C)` to `nodeToPrevNode`, don't add `B` to `queue` as already visited
  - Process `D` in `localQueue`, add `(D->C` to `nodeToPrevNode`, terminate

Now there is no link from `B->A`, instead `B->C` causes an infinite loop. Is that right?

Reply:
Thanks for the counterexample! Now only adding a node key to `nodeToPrevNode` iff that node has not been visited yet to guarantee that all insertions to the `nodeToPrevNode` map have unique keys.

Added a unit test in SchemaTest.java for this method on a very similar graph structure to above. It did run an infinite loop before I put in the code changes haha